include/superior_mysqlpp/extras/prepared_statement_utils.hpp

```diff
@@ -11,89 +11,127 @@ namespace SuperiorMySqlpp
-        auto generatePreparedStatementImpl(Connection &connection, const std::string &query, void (CallableClass::*)(Args...) volatile) {
-            return connection.makePreparedStatement<ResultBindings<std::remove_cv_t<std::remove_reference_t<Args>>...>, storeResult, validateMode, warnMode, ignoreNullable>(query);
-        }
+        template<template<typename...> class ResultTuple, typename... ResultArgs, template<typename...> class ParamsTuple, typename... ParamsArgs>
+        struct ToPreparedStatement<ResultTuple<ResultArgs...>, ParamsTuple<ParamsArgs...>>
+        {
+            /**
+             * Constructs prepared statement
+             * @param connection Connection handle into database
+             * @param query Query for prepared statement
+             * @tparam storeResult Boolean indicating if results will be in `store` or `use` mode
+             * @tparam validateMode Indicates validate mode level
+             * @tparam warnMode Indicates warning mode level
+             * @tparam ignoreNullable Disables null type checking
+             */
+            template <bool storeResult, ValidateMetadataMode validateMode, ValidateMetadataMode warnMode, bool ignoreNullable>
+            static inline auto generate(Connection &connection, const std::string &query)
+            {
+                return PreparedStatement<ResultBindings<ResultArgs...>, ParamBindings<ParamsArgs...>, storeResult, validateMode, warnMode, ignoreNullable>
+                    { connection, query };
+            }
+        };
 
         /**
-         * @brief Makes prepared statement by deducing ResultBindings template arguments from function's signature
+         * @brief Makes prepared statement by deducing ResultBindings and ParamBindings template arguments from functions's signature
          * @param connection Connection handle into database
          * @param query Query for prepared statement
          * @tparam storeResult Boolean indicating if results will be in `store` or `use` mode
          * @tparam validateMode Indicates validate mode level
          * @tparam warnMode Indicates warning mode level
          * @tparam ignoreNullable Disables null type checking
          */
-        template <bool storeResult, ValidateMetadataMode validateMode, ValidateMetadataMode warnMode, bool ignoreNullable, typename CallableClass, typename... Args>
-        auto generatePreparedStatementImpl(Connection &connection, const std::string &query, void (CallableClass::*)(Args...) const) {
-            return connection.makePreparedStatement<ResultBindings<std::remove_cv_t<std::remove_reference_t<Args>>...>, storeResult, validateMode, warnMode, ignoreNullable>(query);
+        template <bool storeResult, ValidateMetadataMode validateMode, ValidateMetadataMode warnMode, bool ignoreNullable, typename ResultCallable, typename ParamCallable>
+        auto generatePreparedStatement(Connection &connection, const std::string &query, ResultCallable &&, ParamCallable &&)
```

Comment:
Note that in `psQuery`, we use order `paramCallable, resultCallable`, but in internal functions it's switched. I feel that may bite us later, especially as they aren't distinct type-wise.